pkg/networkpolicy/networkpolicy.go

```diff
@@ -0,0 +1,300 @@
+func (builder *NetworkPolicyBuilder) WithNamespaceIngressRule(
+	namespaceIngressMatchLabels map[string]string,
+	podIngressMatchLabels map[string]string) *NetworkPolicyBuilder {
+	if valid, err := builder.validate(); !valid {
+		builder.errorMsg = err.Error()
+
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating networkPolicy %s with the ingressRule defined", builder.Definition.Name)
+
+	builder.isMutationAllowed("ingress")
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if namespaceIngressMatchLabels == nil && podIngressMatchLabels == nil {
+		glog.V(100).Infof("At least one type of the selector should be defined")
+
+		builder.errorMsg = "The selectorMatchLabel is empty sting"
+
+		return builder
+	}
+
+	if namespaceIngressMatchLabels == nil {
+		glog.V(100).Infof("The namespaceSelector should not be defined")
+	} else {
+
+		if builder.Definition.Spec.Ingress == nil {
+			builder.Definition.Spec.Ingress = make([]netv1.NetworkPolicyIngressRule, 1)
+		}
+
+		if builder.Definition.Spec.Ingress[0].From == nil {
+			builder.Definition.Spec.Ingress[0].From = make([]netv1.NetworkPolicyPeer, 1)
+		}
+
+		addToFrom := netv1.NetworkPolicyPeer{
+			NamespaceSelector: &metav1.LabelSelector{
+				MatchLabels: namespaceIngressMatchLabels,
+			},
+		}
+
+		builder.Definition.Spec.Ingress[0].From = append(builder.Definition.Spec.Ingress[0].From, addToFrom)
+	}
+
+	if podIngressMatchLabels == nil {
+		glog.V(100).Infof("The podSelector should not be defined")
+	} else {
+
+		if builder.Definition.Spec.Ingress == nil {
+			builder.Definition.Spec.Ingress = make([]netv1.NetworkPolicyIngressRule, 1)
+		}
+
+		if builder.Definition.Spec.Ingress[0].From == nil {
+			builder.Definition.Spec.Ingress[0].From = make([]netv1.NetworkPolicyPeer, 1)
+		}
+
+		addToFrom := netv1.NetworkPolicyPeer{
+			PodSelector: &metav1.LabelSelector{
+				MatchLabels: podIngressMatchLabels,
+			},
+		}
+
+		builder.Object.Spec.Ingress[0].From = append(builder.Object.Spec.Ingress[0].From, addToFrom)
+	}
+
+	return builder
+}
+
+// WithPolicyType add policyType to the networkPolicy.
+func (builder *NetworkPolicyBuilder) WithPolicyType(policyType netv1.PolicyType) *NetworkPolicyBuilder {
+	if valid, err := builder.validate(); !valid {
+		builder.errorMsg = err.Error()
+
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating networkPolicy %s with the policyType defined: %d", builder.Definition.Name)
+
+	builder.isMutationAllowed("policyTypes")
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if policyType == "" {
+		glog.V(100).Infof("The policyType value have to be provided")
+
+		builder.errorMsg = "The policyType is empty sting"
+
+		return builder
+	}
+
+	builder.Definition.Spec.PolicyTypes = append(builder.Definition.Spec.PolicyTypes, policyType)
+
+	return builder
+}
+
+// WithPodSelector add podSelector to the networkPolicy.
+func (builder *NetworkPolicyBuilder) WithPodSelector(podSelectorMatchLabels map[string]string) *NetworkPolicyBuilder {
```

Comment:
I will, but I don't think you a right, few things:

- I don't understand why to remove 'return builder' from the "if podSelectorMatchLabels == nil" block and than build another if block for return, if this is only the case where return with error could be executed???
- Why put getting and assigning a value in one line? in case of a fall, this makes it difficult to debug

Reply:
Regarding the first point:
Anyway we have this block if builder.errorMsg != ""  before we actually update struct so having additional return is simply duplication

Regarding the second point. I don't think we need have standalone variable that is used only once.